pkg/controller/controller.go

```diff
@@ -108,25 +106,97 @@ func (sc *Controller) performDataMigration(schedule []*schema.Schedule) error {
 			// Create migration task only if there are intersections
 			if len(intersections) > 0 {
 				key := fmt.Sprintf("%d->%d", oldShardID, newShardID)
-				if _, exists := migrations[key]; !exists {
-					migrations[key] = &MigrationTask{
-						fromShardInfo: shards[0],
-						toShardInfo:   shards[1],
-						ranges:        intersections,
-					}
+				if _, exists := ops[key]; !exists {
+					ops[key] = NewOperator(shards[0], shards[1], intersections, nil)
 				}
 			}
 		}
 	}
-
 	// Consolidate overlapping ranges for each migration
-	for _, task := range migrations {
-		task.ranges = consolidateKeyRanges(task.ranges)
+	for _, op := range ops {
+		op.keyRanges = schema.ConsolidateKeyRanges(op.keyRanges)
+		sc.operators = append(sc.operators, op)
+
+		ctx, cancel := context.WithTimeout(context.Background(), schema.MIGRATION_TIMEOUT)
```

Comment:
I feel like this timeout should be dynamically set according to the size of the data migrated as an improvement

Reply:
true. Put it into our wishlist